### ISSUE TITLE
RMSD perres ignoring ions

### ISSUE THREAD

**Issue by d-beltran** [Amber-MD/pytraj#1580]
Hi,

I just found out rmsd_perres is ignoring some ions (not always all ions). Those ions are an independent residue with 1 unique atom each. I can provide a pair of examples to reproduce the error if needed.

And by the way, rmsd_perres always prints this message although it works fine: "Error: Range::SetRange(None): Range is -1 for None"

¿Any ideas?

Thank you for maintaining pytraj :)

**Comment by hainm** [comment]
>  I can provide a pair of examples to reproduce the error if needed.

@d-beltran  Please. Thanks.

**Comment by d-beltran** [comment]
I attached 2 pdb and 2 xtc trajectory files.

In the first example there is only 1 zinc atom. it is the penultimate residue in the pdb.

Run this code to reproduce the problem:

`import pytraj as pt

pt_trajectory = pt.iterload('test.xtc', 'test.pdb')
rmsds = pt.rmsd_perres(pt_trajectory)

# Print last 3 residues in topology
topology_residues = list(pt_trajectory.topology.residues)
for residue in topology_residues[-3:]:
    print(residue)
    
# Print last 3 residues in the anlysis output
for residue in rmsds[-3:]:
    print(residue.key)`

As you will see, when I load the trajectory and pdb files the zinc ion is there. However the results of rmsd_perres do not include the zinc residue rmsd.

At the second example there are 4 ions: 2 zinc and 2 magnesium. rmsd_perres returns data only for 1 of the magnesium ions.

[tests.zip](https://github.com/Amber-MD/pytraj/files/6591946/tests.zip)

**Comment by hainm** [comment]
hi @d-beltran: I will look at the issue but I don't think I can resolve this soon. Just want to let you know. Thanks for your files. Cheers.

**Comment by d-beltran** [comment]
Sure. No problem.
Thank you @hainm :)

**Comment by d-beltran** [comment]
Hi again,

Today I had the same problem with a coarse grain simulation.
However, this time there were no results for any residue at all.

I discovered that if I pass the "resrange" argument then all residues are included in the results, also with the two examples above. So in order to avoid this problem this is how the function above should be called:

```python
import pytraj as pt

pt_trajectory = pt.iterload('test.xtc', 'test.pdb')
last_residue_number = pt_trajectory.topology.n_residues + 1
rmsds = pt.rmsd_perres(pt_trajectory, resrange = f'1-{last_residue_number}')
```

Another advantage of passing the resrange argument is that the error message "Error: Range::SetRange(None): Range is -1 for None" disappears. This error message was not a problem, because the results were correct anyway, but it is better gone.

Best regards

**Comment by drroe** [comment]
Hi, I must have missed this issue when it was originally posted, sorry!

@d-beltran the reason this happens is if no residue range is specified, the `rms perres` command of `cpptraj` will default to **solute** residues; this was to avoid situations like calculating the per-residue RMSD of every single water in a fully solvated system. This is noted in the cpptraj manual
```
range <res range> Calculate per-residue RMSDs for residues in <res range> (default all solute residues).
```
but not in the pytraj docs. So @hainm and I need to do a better job keeping the documentation in sync!

So to sum up, the function is working as intended, but should be documented better. Hope this helps!
